lib/SILGen/SILGenDecl.cpp

```diff
@@ -1162,8 +1161,35 @@ void SILGenFunction::emitPatternBinding(PatternBindingDecl *PBD,
     SILValue alet;
     {
       SILLocation loc(PBD);
+
+      // Currently we don't pass any options here, so just grab a 'nil'
+      CanType RawPointerTy = C.TheRawPointerType;
+      CanType OptRawPointerTy =
+          OptionalType::get(RawPointerTy)->getCanonicalType();
+//      SILValue options = getOptionalNoneValue(
+//          loc, getTypeLowering(OptRawPointerTy));
+
+//      auto options = B.createUncheckedBitCast(
+//          loc, ManagedValue::forUnmanaged(B.createIntegerLiteral(0)),
+////          getType
+//          //          Lowering(OptRawPointerTy));
+//          SILType::getPrimitiveAddressType(OptRawPointerTy));
+        SILValue options = B.createUncheckedReinterpretCast(
+          loc, B.createIntegerLiteral(0),
+//          getType
+          //          Lowering(OptRawPointerTy));
+          SILType::getPrimitiveAddressType(OptRawPointerTy));
+
+//      options->dump();
+
+      fprintf(stderr, "[%s:%d] (%s) === RAW PTR OPT\n", __FILE__, __LINE__, __FUNCTION__);
+      getOptionalNoneValue(loc, getTypeLowering(OptRawPointerTy))->dump();
+//      fprintf(stderr, "[%s:%d] (%s) === RAW PTR DIR \n", __FILE__, __LINE__, __FUNCTION__);
+//      getTypeLowering(OptRawPointerTy).dump();
+
       alet = emitAsyncLetStart(
           loc,
+          options,
```

Comment:
Having a lot of trouble figuring this one out...

I'm trying to get a `null` `SILValue` that will "fit" here, the Swift type declaration of what this ends up calling is:

```
@_silgen_name("swift_asyncLet_start")
public func _asyncLetStart<T>(
  asyncLet: Builtin.RawPointer,
  options: Builtin.RawPointer?,
  operation: @Sendable () async throws -> T
)
```

and the RuntimeFunction

```
FUNCTION(AsyncLetStart,
         swift_asyncLet_start, SwiftCC,
         ConcurrencyAvailability,
         RETURNS(VoidTy),
         ARGS(SwiftAsyncLetPtrTy,         // AsyncLet*, alias for Int8PtrTy
              SwiftTaskOptionRecordPtrTy, // options
              TypeMetadataPtrTy,          // futureResultType
              Int8PtrTy,                  // closureEntry
              OpaquePtrTy,                // closureContext
         ),
         ATTRS(NoUnwind, ArgMemOnly))
```

where   SwiftTaskOptionRecordPtrTy is just

```
  SwiftTaskOptionRecordPtrTy = Int8PtrTy; // we pass it opaquely (TaskOptionRecord*)
```

So I need to "make" a `null` of `i8*` in SIL, and get that as a SILValue, we're all happy then...

The various attempts are commented here... in general It seems an unchecked cast of a 0 literal would be right but that ends up in a nasty crash now

```

0.	Program arguments: /Users/ktoso/code/swift-project/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/bin/swift-frontend -frontend -c -primary-file /Users/ktoso/code/swift-project/swift/test/Concurrency/Runtime/async_task_cancellation_early.swift -target x86_64-apple-macosx10.9 -enable-objc-interop -sdk /Applications/Xcode-Internal.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk -F /Applications/Xcode-Internal.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks -F /Users/ktoso/code/swift-project/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/lib -module-cache-path /Users/ktoso/code/swift-project/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/swift-test-results/x86_64-apple-macosx10.9/clang-module-cache -swift-version 4 -define-availability "SwiftStdlib 5.5:macOS 9999, iOS 9999, watchOS 9999, tvOS 9999" -enable-experimental-concurrency -target-sdk-version 11.3 -parse-as-library -module-name main -o /var/folders/w1/hmg_v8p532d800g08jtqtddc0000gn/T/async_task_cancellation_early-dd56ee.o
1.	Swift version 5.5-dev (LLVM b12ac9fa38462e7, Swift 038af80a880ce85)
2.	Compiling with effective version 4.1.50
3.	While evaluating request ASTLoweringRequest(Lowering AST to SIL for file "/Users/ktoso/code/swift-project/swift/test/Concurrency/Runtime/async_task_cancellation_early.swift")
4.	While silgen emitFunction SIL function "@$s4main30test_detach_cancel_child_earlyyyYaF".
 for 'test_detach_cancel_child_early()' (at /Users/ktoso/code/swift-project/swift/test/Concurrency/Runtime/async_task_cancellation_early.swift:14:1)
5.	While silgen closureexpr SIL function "@$s4main30test_detach_cancel_child_earlyyyYaFSbyYaYbcfU_".
 for expression at [/Users/ktoso/code/swift-project/swift/test/Concurrency/Runtime/async_task_cancellation_early.swift:16:44 - line:27:3] RangeText="{
    async let childCancelled: Bool = { () -> Bool in
      await Task.sleep(2_000_000_000)
      return Task.isCancelled
    }()

    let xx = await childCancelled
    print("child, cancelled: \(xx)") // CHECK: child, cancelled: true
    let cancelled =  Task.isCancelled
    print("self, cancelled: \(cancelled )") // CHECK: self, cancelled: true
    return cancelled
  "
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend           0x0000000109620a67 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 39
1  swift-frontend           0x000000010961fc15 llvm::sys::RunSignalHandlers() + 85
2  swift-frontend           0x0000000109621096 SignalHandler(int) + 262
3  libsystem_platform.dylib 0x00007fff204b7d7d _sigtramp + 29
4  libsystem_platform.dylib 0x00007ffeeb4c8c90 _sigtramp + 18446744072820428592
5  swift-frontend           0x00000001049b9383 swift::SILBuilder::createIntegerLiteral(swift::IntegerLiteralExpr*) + 131
6  swift-frontend           0x0000000104f07213 swift::Lowering::SILGenFunction::emitPatternBinding(swift::PatternBindingDecl*, unsigned int) + 963
7  swift-frontend           0x0000000104f0769d swift::Lowering::SILGenFunction::visitPatternBindingDecl(swift::PatternBindingDecl*) + 45
8  swift-frontend           0x0000000104f82e1d swift::ASTVisitor<(anonymous namespace)::StmtEmitter, void, void, void, void, void, void>::visit(swift::Stmt*) + 429
9  swift-frontend           0x0000000104f82c65 swift::Lowering::SILGenFunction::emitStmt(swift::Stmt*) + 21
10 swift-frontend           0x0000000104f3599e swift::Lowering::SILGenFunction::emitClosure(swift::AbstractClosureExpr*) + 430
11 swift-frontend           0x0000000104eb4a9a swift::Lowering::SILGenModule::emitFunctionDefinition(swift::SILDeclRef, swift::SILFunction*) + 1706
12 swift-frontend           0x0000000104eb7c16 swift::Lowering::SILGenModule::emitClosure(swift::AbstractClosureExpr*) + 134
13 swift-frontend           0x0000000104f2bb3c (anonymous namespace)::RValueEmitter::visitAbstractClosureExpr(swift::AbstractClosureExpr*, swift::Lowering::SGFContext) + 124
14 swift-frontend           0x0000000104f1f730 swift::ASTVisitor<(anonymous namespace)::RValueEmitter, swift::Lowering::RValue, void, void, void, void, void, swift::Lowering::SGFContext>::visit(swift::Expr*, swift::Lowering::SGFContext) + 128
15 swift-frontend           0x0000000104f16357 swift::Lowering::SILGenFunction::emitRValueAsSingleValue(swift::Expr*, swift::Lowering::SGFContext) + 55
16 swift-frontend           0x0000000104f254ab swift::ASTVisitor<(anonymous namespace)::RValueEmitter, swift::Lowering::RValue, void, void, void, void, void, swift::Lowering::SGFContext>::visit(swift::Expr*, swift::Lowering::SGFContext) + 24059
17 swift-frontend           0x0000000104f16357 swift::Lowering::SILGenFunction::emitRValueAsSingleValue(swift::Expr*, swift::Lowering::SGFContext) + 55
18 swift-frontend           0x0000000104e9d2c0 swift::Lowering::ArgumentSource::getAsSingleValue(swift::Lowering::SILGenFunction&, swift::Lowering::SGFContext) && + 352
19 swift-frontend           0x0000000104f03405 swift::Lowering::SILGenFunction::emitConvertedRValue(swift::SILLocation, swift::Lowering::Conversion const&, swift::Lowering::SGFContext, llvm::function_ref<swift::Lowering::ManagedValue (swift::Lowering::SILGenFunction&, swift::SILLocation, swift::Lowering::SGFContext)>) + 357
20 swift-frontend           0x0000000104e9d438 swift::Lowering::ArgumentSource::getConverted(swift::Lowering::SILGenFunction&, swift::Lowering::Conversion const&, swift::Lowering::SGFContext) && + 120
21 swift-frontend           0x0000000104ed4cbe (anonymous namespace)::ArgEmitter::emit(swift::Lowering::ArgumentSource&&, swift::Lowering::AbstractionPattern) + 2686
22 swift-frontend           0x0000000104ec3177 (anonymous namespace)::ArgEmitter::emitSingleArg(swift::Lowering::ArgumentSource&&, swift::Lowering::AbstractionPattern) + 679
23 swift-frontend           0x0000000104ed3de3 (anonymous namespace)::ArgEmitter::emitPreparedArgs(swift::Lowering::PreparedArguments&&, swift::Lowering::AbstractionPattern) + 147
24 swift-frontend           0x0000000104edd2de (anonymous namespace)::CallSite::emit(swift::Lowering::SILGenFunction&, swift::Lowering::AbstractionPattern, swift::CanTypeWrapper<swift::SILFunctionType>, (anonymous namespace)::ParamLowering&, llvm::SmallVectorImpl<swift::Lowering::ManagedValue>&, llvm::SmallVectorImpl<(anonymous namespace)::DelayedArgument>&, swift::Lowering::CalleeTypeInfo::ForeignInfo const&) && + 926
25 swift-frontend           0x0000000104edca6f (anonymous namespace)::CallEmission::emitArgumentsForNormalApply(swift::Lowering::AbstractionPattern, swift::CanTypeWrapper<swift::SILFunctionType>, swift::Lowering::CalleeTypeInfo::ForeignInfo const&, llvm::SmallVectorImpl<swift::Lowering::ManagedValue>&, llvm::Optional<swift::SILLocation>&) + 1631
26 swift-frontend           0x0000000104ec744a (anonymous namespace)::CallEmission::apply(swift::Lowering::SGFContext) + 2794
27 swift-frontend           0x0000000104ec4843 swift::Lowering::SILGenFunction::emitApplyExpr(swift::ApplyExpr*, swift::Lowering::SGFContext) + 51
28 swift-frontend           0x0000000104f1f716 swift::ASTVisitor<(anonymous namespace)::RValueEmitter, swift::Lowering::RValue, void, void, void, void, void, swift::Lowering::SGFContext>::visit(swift::Expr*, swift::Lowering::SGFContext) + 102
29 swift-frontend           0x0000000104f13b62 swift::Lowering::SILGenFunction::emitExprInto(swift::Expr*, swift::Lowering::Initialization*, llvm::Optional<swift::SILLocation>) + 258
30 swift-frontend           0x0000000104f07474 swift::Lowering::SILGenFunction::emitPatternBinding(swift::PatternBindingDecl*, unsigned int) + 1572
31 swift-frontend           0x0000000104f0769d swift::Lowering::SILGenFunction::visitPatternBindingDecl(swift::PatternBindingDecl*) + 45
32 swift-frontend           0x0000000104f82e1d swift::ASTVisitor<(anonymous namespace)::StmtEmitter, void, void, void, void, void, void>::visit(swift::Stmt*) + 429
33 swift-frontend           0x0000000104f82c65 swift::Lowering::SILGenFunction::emitStmt(swift::Stmt*) + 21
34 swift-frontend           0x0000000104f3517a swift::Lowering::SILGenFunction::emitFunction(swift::FuncDecl*) + 362
35 swift-frontend           0x0000000104eb5c3e swift::Lowering::SILGenModule::emitFunctionDefinition(swift::SILDeclRef, swift::SILFunction*) + 6222
36 swift-frontend           0x0000000104eb7887 emitOrDelayFunction(swift::Lowering::SILGenModule&, swift::SILDeclRef, bool) + 231
37 swift-frontend           0x0000000104eb43ce swift::Lowering::SILGenModule::emitFunction(swift::FuncDecl*) + 126
38 swift-frontend           0x0000000104eb9ea4 swift::ASTLoweringRequest::evaluate(swift::Evaluator&, swift::ASTLoweringDescriptor) const + 1988
39 swift-frontend           0x0000000104f827c0 swift::SimpleRequest<swift::ASTLoweringRequest, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> > (swift::ASTLoweringDescriptor), (swift::RequestFlags)9>::evaluateRequest(swift::ASTLoweringRequest const&, swift::Evaluator&) + 128
40 swift-frontend           0x0000000104ebdcb3 llvm::Expected<swift::ASTLoweringRequest::OutputType> swift::Evaluator::getResultUncached<swift::ASTLoweringRequest>(swift::ASTLoweringRequest const&) + 403
41 swift-frontend           0x0000000104ebae01 swift::performASTLowering(swift::FileUnit&, swift::Lowering::TypeConverter&, swift::SILOptions const&) + 113
42 swift-frontend           0x000000010480a1d2 performCompileStepsPostSema(swift::CompilerInstance&, int&, swift::FrontendObserver*) + 226
43 swift-frontend           0x00000001047fe7a6 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 4838
44 swift-frontend           0x0000000104751212 main + 514
45 libdyld.dylib            0x00007fff2048df3d start + 1
<unknown>:0: error: unable to execute command: Segmentation fault: 11
<unknown>:0: error: compile command failed due to signal 11 (use -v to see invocation)
```